### ISSUE TITLE
yolov8 Obb low confidence

### ISSUE THREAD

**Issue by zkq-afk** [ultralytics/ultralytics#24541]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I have a problem. When training YOLOv8-OBB, I used more than 20,000 images with about 120 classes of labels. I trained the model based on the YOLOv8x-OBB backbone for 300 epochs. After training, when performing inference on over 900 images with a confidence threshold of 0.41, about 40 images had missing bounding boxes. However, all objects were detected when the confidence threshold was set to 0.02, but I noticed the confidence scores were very low. What methods can I use to improve the confidence scores?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zkq-afk, thank you for your interest in Ultralytics 🚀! This is an automated response to help gather the details our team needs, and an Ultralytics engineer will also assist you soon.
We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python) and [CLI](https://docs.ultralytics.com/usage/cli) usage examples and where many of the most common questions may already be answered.
For this custom training ❓ question about OBB confidence scores, please provide as much information as possible, including dataset image examples, validation results, training logs, your exact train/val/predict commands, and confirmation that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips).
If you believe there may also be a 🐛 bug involved, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.
Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.
## Upgrade
Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:
```bash
pip install -U ultralytics
```
## Environments
YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>
## Status
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes) and [Tasks](https://docs.ultralytics.com/tasks) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by zkq-afk** [comment]
epochs=100, time=None, patience=50, batch=4, imgsz=640, save=True, save_period=-1, cache=False, device=[0], workers=0, project=None, name=frontv11Xobb_reluall42, exist_ok=False, pretrained=True, optimizer=SGD, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, val_period=25, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=True, opset=None, workspace=None, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=1.5, dfl=1.5, pose=12.0, kobj=1.5, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.5, fliplr=0.5, bgr=0.0, mosaic=0.95, mixup=0.1, copy_paste=0.0, copy_paste_mode=flip, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml
These are my experimental parameters. This model is applied to industrial inspection scenarios, where each image contains 10 to 30 objects to be detected.

**Comment by colchicinewf** [comment]
I also encountered the same issue,When training yolov26,the model's precision is relatively high. When I tried to lower the confidence threshold to reduce precision and improve recall,I found that the threshold was already at 0.01 and couldn't be lowered any futher.

**Comment by tedo001** [comment]
bruh use a low threshold range like 0.25 to 0.35 to encounter this problem bro or else change your module into yolo26 variants like a yolo26n or yolo26x.......

**Comment by raimbekovm** [comment]
thanks @zkq-afk. with 20k images across 120 classes you average ~166 imgs/class, well below ultralytics' [recommended ≥1500 imgs and ≥10000 instances per class](https://docs.ultralytics.com/yolov5/tutorials/tips_for_best_training_results/); "many detections in 0.02-0.4 with few above 0.41" is the typical undertraining / per-class data scarcity signature.

the single most useful thing to do first is run `model.val()` and look at `metrics.box.maps`, `p`, `r` **per class** (see the [performance metrics guide](https://docs.ultralytics.com/guides/yolo-performance-metrics/)) - your 40 failing images almost certainly concentrate on a few underrepresented classes, and that tells you exactly where to add data, oversample, weight, or merge visually similar classes (the [weighted dataloader tutorial](https://y-t-g.github.io/tutorials/yolo-class-balancing/) is the canonical pattern). beyond that, the highest-impact knobs are raising resolution (`imgsz=1024`, or tiling 1280→640 per the [evaluation insights guide](https://docs.ultralytics.com/guides/model-evaluation-insights/)) for small industrial details, and training longer - your `epochs=100 patience=50` very likely stops just as rare classes start converging (docs default is 300 epochs). for a new run, `yolo26x-obb.pt` is the newer generation (NMS-free, MuSGD) and a reasonable upgrade from v8x-obb.

a couple of non-levers worth knowing: `predict(augment=True)` does not work for OBB - it falls back to single-scale with a warning (`tasks.py:457-459`), so TTA is not available for OBB models. and aggressive `degrees=180` rotation aug on top of a pretrained obb checkpoint did not help in a controlled A/B i ran on dota128 (mAP50 -0.003, mean conf of detections>0.41 went 0.795 → 0.755 vs degrees=0); it may still help when training from scratch on a narrow-rotation domain, but it isn't a guaranteed lever. per-class numbers are the most informative thing to share if you want to dig further.

confidence calibration depends on a lot of dataset- and domain-specific factors (class similarity, label noise, object scale distribution, intra-class variance, etc.), so the exact gains will vary, but these are the patterns that usually move the needle on this symptom.

**Comment by zkq-afk** [comment]
Sorry, I didn't quite get what you meant. But my dataset is actually pretty large, since each of my images contains a great many objects covering all kinds of categories. I’ve checked those over 40 images with missed detections, and there’s no common feature among the overlooked objects — they don’t belong to specific fixed categories at all, which is what confuses me the most. If the problem lay with insufficient category data, I could just add more training samples to fix it.
Did you mean I should increase the training epochs to 300 and check the results? I’d also like to ask what other methods I can use to boost confidence scores. I used to have far more missed detections, around 80 problematic images in total. I managed to fix part of the issue by setting the cls threshold to 1.5, yet raising it further to 2.5 brought no obvious improvement.

**Comment by glenn-jocher** [comment]
`cls=1.5` is the training classification loss gain, not an inference threshold, so increasing it will not reliably raise `results[0].obb.conf`; yes, a clean `epochs=300` run is worth trying, but for dense industrial OBB scenes the biggest practical levers are usually higher `imgsz` (for example `1024`) and less aggressive augmentation, especially very high `mosaic`/`mixup`, then compare the `model.val()` text output to see whether the weakness is class-, size-, or image-specific. See the [OBB `conf` reference](https://docs.ultralytics.com/reference/engine/results#ultralytics.engine.results.OBB.conf) and the [confidence overview](https://www.ultralytics.com/glossary/confidence).

**Comment by onuralpszr** [comment]
@lmycross can you take look here please ?

**Comment by lmycross** [comment]
Hi @zkq-afk, based on your config, I think the low confidence is probably not mainly caused by insufficient epochs. There are several settings that are more likely to affect confidence.

The first thing I would change is `imgsz`. You trained with `imgsz=640`, but your images contain many objects. If many objects are small or crowded, 640 may not preserve enough detail, which can directly reduce box quality and classification confidence. I would recommend 1024 first, and 1280 if GPU memory allows.

The second thing is batch size. Your batch size is only 4. For YOLOv8x-OBB this may be due to GPU memory, but small batch training can be unstable. If possible, use `batch=-1` or the largest batch your GPU supports.

About epochs: your current `epochs=100` may be a bit short for 20k images and 120 classes. I would try 150 to 200 epochs, but only after changing `imgsz` and batch size. Simply training longer with the current settings may not solve the low-confidence issue.

**Comment by zkq-afk** [comment]
Would switching from OBB to HBB yield better results? Could the extra rotation angle parameter of OBB be the cause of the performance impact?

**Comment by lmycross** [comment]
Converting OBB labels to HBB and training a normal detection model can be a useful experiment, but I would treat it as an ablation test, not a guaranteed solution.
HBB training is usually easier than OBB because the model only needs to learn axis-aligned boxes. If your objects do not require precise rotation angles, then HBB may give higher confidence scores and more stable detections.
